### PR TITLE
Move UnwindSafe, RefUnwindSafe, AssertUnwindSafe to core

### PR DESCRIPTION
**Pull request by dtolnay**
They were previously only available in std::panic, not core::panic.

- https://doc.rust-lang.org/1.51.0/std/panic/trait.UnwindSafe.html
- https://doc.rust-lang.org/1.51.0/std/panic/trait.RefUnwindSafe.html
- https://doc.rust-lang.org/1.51.0/std/panic/struct.AssertUnwindSafe.html

Where this is relevant: trait objects! Inside a `#![no_std]` library it's otherwise impossible to have a struct holding a trait object, and at the same time can be used from downstream std crates in a way that doesn't interfere with catch_unwind.

```rust
// common library

#![no_std]

pub struct Thing {
    pub(crate) x: &'static (dyn SomeTrait + Send + Sync),
}

pub(crate) trait SomeTrait {...}
```

```rust
// downstream application

fn main() {
    let thing: library::Thing = ...;
    let _ = std::panic::catch_unwind(|| { let _ = thing; });  // does not work :(
}
```

See https://github.com/dtolnay/colorous/blob/a4131708e2f05d2377964981896ff62dbc9b027b/src/gradient.rs#L7-L15 for a real life example of needing to work around this problem. In particular that workaround would not even be viable if implementors of the trait were provided externally by a caller, as the `feature = "std"` would become non-additive in that case.

What happens without the UnwindSafe constraints:

```rust
fn main() {
    let gradient = colorous::VIRIDIS;
    let _ = std::panic::catch_unwind(|| { let _ = gradient; });
}
```

```console
error[E0277]: the type `(dyn colorous::gradient::EvalGradient + Send + Sync + 'static)` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
   --> src/main.rs:3:13
    |
3   |     let _ = std::panic::catch_unwind(|| { let _ = gradient; });
    |             ^^^^^^^^^^^^^^^^^^^^^^^^ `(dyn colorous::gradient::EvalGradient + Send + Sync + 'static)` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
    | 
   ::: .rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:430:40
    |
430 | pub fn catch_unwind<F: FnOnce() -> R + UnwindSafe, R>(f: F) -> Result<R> {
    |                                        ---------- required by this bound in `catch_unwind`
    |
    = help: within `Gradient`, the trait `RefUnwindSafe` is not implemented for `(dyn colorous::gradient::EvalGradient + Send + Sync + 'static)`
    = note: required because it appears within the type `&'static (dyn colorous::gradient::EvalGradient + Send + Sync + 'static)`
    = note: required because it appears within the type `Gradient`
    = note: required because of the requirements on the impl of `UnwindSafe` for `&Gradient`
    = note: required because it appears within the type `[closure@src/main.rs:3:38: 3:62]`
```

### CODE REVIEW

**Comment by dtolnay** [comment]
- Rebased to resolve conflict with #84730 in library/std/src/panic.rs.

**Comment by dtolnay** [comment]
We discussed this change briefly in a meeting of the libs team. One concern was that there is a sense we may want to deprecate `UnwindSafe` and `RefUnwindSafe` and remove them as a bound from the `catch_unwind` argument. (Probably this would require an RFC, to get better visibility from people who may currently be finding the bound valuable. There is no way to add it back later!)

From following up with @Amanieu, whenever that deprecation were to happen it seems fine to deprecate simultaneously in both places (core::panic and std::panic). In the meantime I think I'd still like to have the traits available for using in trait objects in no_std crates.

Gonna try this to get an impression from folks who weren't in that meeting too:
@rfcbot fcp merge
@rust-lang/libs

**Comment by joshtriplett** [comment]
I would like to see them deprecated as well, but I don't see a good reason to delay making them available from core in the interim.

**Comment by dtolnay** [comment]
- Rebase to resolve conflict with #86603. @joshtriplett

**Comment by dtolnay** [comment]
- Rebase to resolve conflict with #86726

**Comment by dtolnay** [comment]
- Rebase to resolve conflict with #87445. @joshtriplett

File: library/alloc/src/panic.rs [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+use crate::sync::Arc;
+use core::panic::{RefUnwindSafe, UnwindSafe};
+
+// not covered via the Shared impl in libcore b/c the inner contents use
+// Cell/AtomicUsize, but the usage here is unwind safe so we can lift the
+// impl up one level to Arc/Rc itself
+#[stable(feature = "catch_unwind", since = "1.9.0")]
+impl<T: RefUnwindSafe + ?Sized> UnwindSafe for Rc<T> {}
+#[stable(feature = "catch_unwind", since = "1.9.0")]
+impl<T: RefUnwindSafe + ?Sized> UnwindSafe for Arc<T> {}
```
Comment (Amanieu) [inline library/alloc/src/panic.rs:11]: These impls could just be in the `rc`/`sync` modules directly rather than in this file.

Comment (dtolnay) [reply library/alloc/src/panic.rs]: :+1: Moved the impls.

**Pull request merged**